modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/leases/LeaseBatchSerializer.java

```diff
@@ -0,0 +1,485 @@
+import org.apache.ignite.internal.util.io.IgniteDataOutput;
+import org.apache.ignite.internal.versioned.VersionedSerializer;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * {@link VersionedSerializer} for {@link LeaseBatch} instances.
+ *
+ * <p>The following optimizations are applied to minimize the amount of bytes a batch is serialized to:</p>
+ * <ul>
+ *     <li>Java Serialization is not used to serialize components (it's pretty verbose)</li>
+ *     <li>Varints are used extensively</li>
+ *     <li>A dictionary of all nodes mentioned as lease holders and proposed candidates is collected and written in the header once
+ *     per batch. This is beneficial as we usually a lot more leases than number of nodes in cluster, so we can just represent nodes
+ *     and their names as indices in the dictionary (this is especially effective given we use varints as indices are usually very small).
+ *     </li>
+ *     <li>Leases are grouped per table/zone ID (aka object ID), so an object ID is only written once per table </li>
+ * </ul>
+ */
+public class LeaseBatchSerializer extends VersionedSerializer<LeaseBatch> {
+    /*
+     * The following optimizations are applied to minimize the number of bytes a batch is serialized to:
+     *
+     * - Java Serialization is not used to serialize components (it's pretty verbose)
+     * - Varints are used extensively
+     * - A dictionary of all nodes mentioned as lease holders and proposed candidates is collected and written in the header once
+     * per batch. This is beneficial as we usually a lot more leases than number of nodes in cluster, so we can just represent nodes
+     * and their names as indices in the dictionary (this is especially effective given we use varints as indices are usually very small).
+     * - Leases are grouped per table/zone ID (aka object ID), so an object ID is only written once per object
+     * - Object IDs are represented as deltas with respect to the previous object; as leases are sorted by object IDs, the deltas are
+     * encoded effectively as varints
+     * - Partition IDs are not written, instead their index represents partition ID. Holes are rare (if possible), so we just write
+     * special markers to denote holes
+     * - Expiration timestamps are always computed from Start timestamps. To account for a possibility of this duration being changed
+     * on the go, we pass the most frequent duration once per batch (in the header), and then we only write lease duration for leases with
+     * uncommon duration (that is, duration different from the most frequent one). We use a flag to distinguish such leases.
+     * - Expiration timestamps always have their logical part equal to 0, so it's never written.
+     * - Physical parts of Start timestamps are coded as deltas with respect to the least timestamp present in the batch; this allows to
+     * have those deltas small and encode them effectively as varints
+     */
```

Comment:
why not put it all to the javadoc?

Reply:
Removed this block as it duplicates what we have in Javadoc